library/Vanilla/Embeds/AbstractEmbed.php

```diff
@@ -0,0 +1,274 @@
+<?php
+/**
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license GPLv2
+ */
+
+namespace Vanilla\Embeds;
+
+use Exception;
+use Garden\Http\HttpRequest;
+use Garden\Http\HttpResponse;
+use Vanilla\InjectableInterface;
+
+/**
+ * Basic embed class.
+ */
+abstract class AbstractEmbed implements InjectableInterface {
+
+    /** @var bool Allow access to local resources? */
+    private $allowLocal = false;
+
+    /** @var HttpRequest HTTP request interface. */
+    private $httpRequest;
+
+    /** @var string[] Valid domains for this embed type. */
+    protected $domains = [];
+
+    /** @var bool Allow network requests (e.g. HTTP)? */
+    private $networkEnabled = true;
+
+    /** @var string Rendered type of this embed (e.g. video, image). */
+    protected $renderType;
+
+    /** @var string[] Client-side scripts for handling this embed type.  */
+    protected $scripts = [];
+
+    /** @var string Primary type for this embed (e.g. youtube, twitter). */
+    protected $type;
+
```

Comment:
You should have the following constructor in this class:

```
public function __construct(string $type, string $renderType) {
    $this->type = $type;
    $this->renderType = $renderType;
}
```

This will ensure that these 2 uninitialized variables get properly set by extending classes.

A good example of why this is important is [this](https://github.com/vanilla/vanilla/pull/7075/files#diff-0611fab7a20bc99c64bb81aa7380b5c8R12)

$renderType is not defined in that class.
You are also "masking" the super class `$type` variable which is not something that *should* not be done. It does work as intended since PHP doesn't support "down casting" but it is still a bad practice.


Doing `(new \Vanilla\Embeds\ImageEmbed())->getRenderType();` results in 
```
fatal Error in Vanilla\Embeds\AbstractEmbed.getRenderType();
Return value of Vanilla\Embeds\AbstractEmbed::getRenderType() must be of the type string, null returned
The error occurred on or near: /srv/vanilla-repositories/vanilla/library/Vanilla/Embeds/AbstractEmbed.php
59:      * Get the render type for this embed (e.g. video, image).
60:      * @return string
61:      */
62:     public function getRenderType(): string {
63:         return $this->renderType;
64:     }
65: 
66:     /**
67:      * Get scripts for this embed type.

```